Set Content-Security-Policy and Cache-Control Headers

taylorsilva opened the pull request:
## What does this PR accomplish?
Bug Fix | **Feature** | Documentation

## Changes proposed by this PR:

- Add CSP headers that are configurable. Defaults to `frame-ancestors 'none'`
- Add a default cache-control header set to `no-store, private`

## Notes to reviewer:
~I'm not 100% sure about the cache-control values, but looking at the
header of some big sites (i.e. stackoverflow) I found they do set this header
to `private`.~

## Release Note

* A Content-Security-Policy header is now set with a default value that will block framing of the Concourse web UI. This was already possible with the default value of the X-Frames-Option header.
  * The CSP header value is configurable with `CONCOURSE_CONTENT_SECURITY_POLICY`
* A Cache-Control header is set on every page with a default value of `no-store, private`. The value of the header is overwritten for some paths (i.e. web assets)

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [x] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

taylorsilva committed:
Add CSP headers

defaults to `frame-ancestors 'none'` to block framing. Users can
configure the header to have a more robust CSP to meet their security
needs.

Docs: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

taylorsilva committed:
Set default cache-control headers

I'm not 100% sure this is the best thing to do, but looking at the
header of some big sites (i.e. stackoverflow) I found they do set this header
to `private`.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw reviewed (changes requested):
@EstebanFS and I tried to better understand what these headers do. It sounds like the `frame-ancestors` in `Content-Security-Policy` obsoletes `X-Frame-Options`, which we also allow configuring - however, it isn't supported in IE, so probably makes sense to support both.

For `Cache-Control`, I guess these settings just mean don't cache anything. This makes sense for things that should be private/sensitive (like API responses), but probably isn't the best for things like assets (icons, stylesheets, javascript).

Also interesting that we explicitly set `Cache-Control` to cache these assets: https://github.com/concourse/concourse/blob/master/web/cache_handler.go. It seems like both values are sent back in the response - not sure which takes precedence

<img width="281" alt="Screen Shot 2021-04-29 at 5 30 16 PM" src="https://user-images.githubusercontent.com/26583442/116621013-9340cb00-a910-11eb-90b8-672cfa56dc4a.png">

I feel like a default value for `Cache-Control` of `private` might be a bit better, but not sure if that'll satisfy the vulnerability scanner.

analytically commented:
I'll be adding support for HSTS (https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security) in a separate PR once this one is merged.

taylorsilva commented:
Interesting about both Cache-Control headers being returned. I thought later handlers would overwrite it. Not sure how browsers are meant to handle conflicting headers like that either. We can probably write smarter handler/wrapper so hard assets are cached and everything else is not.

Also thought about setting it to `private` instead of `private, no-store`. Didn't for the reason you gave, not sure if it will satisfy the vulnerability scanner :(

taylorsilva committed:
Set Cache-Control to private

removed no-store

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw commented:
> I thought later handlers would overwrite it

I think we can get it to overwrite by using `w.Header.Set` rather than `w.Handler.Add` (in `web/cache_handler`): https://github.com/concourse/concourse/blob/63c757c1a8e8dc7054c0b18d3110c24540756b78/web/cache_handler.go#L14

taylorsilva committed:
Have cache_handler overwrite any previous cache-control headers

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

taylorsilva commented:
I changed Cache-Control to only have `private`. Also changed the cache_handler to overwrite any previous `cache-control` headers as per your comment. I checked git blame history on that line but didn't see any reason why it **had** to be `Add` instead of `Set`, and I can't think of any myself. Didn't break any unit tests either.

aoldershaw commented on atc/wrappa/security_handler.go:
```diff
@@ -18,6 +18,6 @@ func (handler SecurityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request)
 	w.Header().Set("X-XSS-Protection", "1; mode=block")
 	w.Header().Set("X-Content-Type-Options", "nosniff")
 	w.Header().Set("X-Download-Options", "noopen")
-	w.Header().Set("Cache-Control", "no-store, private")
```
Actually 😁 could we add `no-store` back here? It does seem like a reasonable thing for API responses, assuming my understanding is correct, and I'm guessing `private` alone won't satisfy the vulnerability scanner if it still stores responses to disk.

Sorry about that 😬  I think I felt that if it was difficult overriding the `Cache-Control` header per route, then we'd be better off still allowing assets to be properly cached - but given that it's easy to cache assets while still preventing API responses from being cached, I think `no-store` is reasonable to keep

taylorsilva replied:
Good point! I'll add it back in.

taylorsilva committed:
Revert "Set Cache-Control to private"

This reverts commit fc320c0885f766e6b069d3cc804a963b93baa114.

aoldershaw reviewed (approved):
LGTM! Note to self - we'll need to update CI to allow embedding iframes like we do with `XFrameOptions`: https://github.com/concourse/infrastructure/blob/dbf08cf0f07856beb2396c851edb531a1ed1f50b/terraform/environments/production/ci-values.yml.tpl#L14

taylorsilva commented:
I'm gonna go make the packaging PR's for this now.
